text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+
+It'd be great if we had scoped trait impls, for example, so we could do that
+in a way where it's up to the trait author how visible things get.  But that's
+a *massive* feature, so it would be nice not to block on it.
+
+Or libs-api could create a new trait and a new `derive` that's implemented using
+the same magic that today's `derive(PartialOrd)` uses.  But that's another big
+bikeshed, and doesn't even work very well for the "I'm writing my own customized
+derive" cases that just want to use the discriminant internally.
+
+The goal here is to do something easy using syntactic space that's not particularly
+valuable anyway -- if people end up almost never using this directly because there's
+a popular community `derive`, that's great.
+
+## What about `as`?
+
+While `as` *works* on field-less enums, it's not that great there either.
+
+It has the fundamental problem that you have to write out the target type that you want,
+and the wrong one will silently truncate.  This hits the same general "`as` is error-prone"
+theme that is pushing people away from using `as` to using more-specific things
+instead that are either lossless or clearer, to help avoid mistakes.
+
+If this exists, I wouldn't be surprised to see people using `foo.enum#discriminant`
+even in places where `foo as u8` works and is shorter since you don't have to think
+"what was the `repr` of this, again?" and you just get the right thing.
+
+Should the enum's declared `repr` not be the type you actually want, you can always
+use `.enum#discriminant` and *then* `as` cast it -- or hopefully `.into()` or
+something else with clearer intent -- into the type you need.
+
+# Prior art
+[prior-art]: #prior-art
+
+C++'s `std::variant` has an [`index`](https://en.cppreference.com/w/cpp/utility/variant/index)
+method, which always returns `std::size_t` since there's no custom discriminants.
+(It's more like what rustc calls a *variant index* internally.)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Is auto-deref worth it?  I would propose leaving it in the RFC for merging,
+  as wanting to use this on `&Enum` will be common, but if in the course of
+  implementing it's particularly annoying then stabilizing without it would
+  be tolerable, since error messages could suggest the correct thing.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+If this turns out to work well, there's a variety of related properties of things
+which could be added in ways similar to this.
+
+For example, you could imagine `MyEnum::enum#VARIANT_COUNT` saying how many variants
+are declared, `MyEnum::enum#ReprType` to get the type of the discriminant, or
+`my_enum.enum#variant_index` to get the declaration-order index of the variant
+(as opposed to its *discriminant* value).
+
+Those are *much* easier to generate with a proc macro, however, so are not included
+in this RFC.  They would need separate motivation from what's done here.
```

Comment:
N.B.: The discriminant type is actually outright impossible to reflect upon syntactically in a proc macro, because the discriminant type of `repr(C)` enums is platform-specific.

This both warrants a correction in the RFC (it's not much easier to generate), and should be cited as a motivating reason why this RFC's proposal is quite useful! Even though it's not *too* difficult to syntactically [reflect on the discriminant values](https://github.com/rust-lang/rfcs/pull/3607/files#r1561342368), it's impossible to syntactically determine the return type of the derived discriminant accessor should be if the enum is `repr(C)`.